cmd/egress-cni-plugin/egressContext.go

```diff
@@ -233,101 +239,426 @@ func (c *EgressContext) SetupHostVeth(vethName string) error {
+			containerInterface.Name, containerIPv6)
+		return fmt.Errorf("no IPv6 global unicast address found, ifName: %s, IPs: %s",
+			containerInterface.Name, containerIPv6)
+	}
+
+	err = ec.setupContainerIPv6Route(hostInterface, containerInterface)
+	if err != nil {
+		ec.Log.Errorf("setupContainerIPv6Route failed: %v", err)
+		return err
+	}
+	ec.Log.Debugf("container route set up successfully")
+
+	err = ec.setupHostIPv6Route(hostInterface, containerIPv6[0])
+	if err != nil {
+		ec.Log.Errorf("setupHostIPv6Route failed: %v", err)
+		return err
+	}
+	ec.Log.Debugf("host IPv6 route set up successfully")
+
+	// set up SNAT in host for container IPv6 egress traffic
+	// following line adds an ip6tables entries to NAT for IPv6 traffic between container v6if0 and node primary ENI (eth0)
+	err = snat.Add(ec.IPTablesIface, ec.NetConf.NodeIP, containerIPv6[0], ipv6MulticastRange, ec.SnatChain, ec.SnatComment, ec.NetConf.RandomizeSNAT)
+	if err != nil {
+		ec.Log.Errorf("setup host snat failed: %v", err)
+		return err
+	}
+
+	ec.Log.Debugf("host IPv6 SNAT set up successfully")
+
+	mergeResult(ec.Result, ec.TmpResult)
+	// remove comment for debug purpose
+	// ec.Log.Debugf("output result: %+v", *ec.Result)
+
+	// Pass through the previous result
+	return types.PrintResult(ec.Result, ec.NetConf.CNIVersion)
+}
+
+// cmdDelEgressV6 exec clear the setting to support IPv6 egress traffic in EKS IPv4 cluster
+func (ec *egressContext) cmdDelEgressV6() (err error) {
+	var contIPAddrs []netlink.Addr
+
+	if ec.IPTablesIface == nil {
+		if ec.IPTablesIface, err = ec.IptCreator(iptables.ProtocolIPv6); err != nil {
+			ec.Log.Error("command ip6tables not found")
+		}
+	}
+
+	if ec.NsPath != "" {
+		err = ec.Ns.WithNetNSPath(ec.NsPath, func(hostNS ns.NetNS) error {
+			var containerLink netlink.Link
+			var _err error
+			ec.Log.Debugf("trying to get link by name %s", ec.NetConf.IfName)
+			containerLink, _err = ec.Link.LinkByName(ec.NetConf.IfName)
+			if _err != nil {
+				if !cniutils.IsLinkNotFoundError(_err) {
+					ec.Log.Debugf("failed to get link by name %s: %v", ec.NetConf.IfName, _err)
+				}
+				// must return when err, following code needs containerLink
+				return _err
 			}
 
-			if err != nil && err == ip.ErrLinkNotFound {
-				c.Log.Debugf("DEL: Link Not Found, returning", err)
-				return nil
+			contIPAddrs, _err = ec.Link.AddrList(containerLink, netlink.FAMILY_V6)
+			if _err != nil {
+				ec.Log.Debugf("failed to get container link %s IPv6 address, host may leave NAT rules uncleared: %v",
+					containerLink.Attrs().Name, _err)
+				return _err
 			}
-			return err
+
+			_err = ec.Link.LinkDel(containerLink)
+			if _err != nil {
+				if !cniutils.IsLinkNotFoundError(_err) {
+					ec.Log.Debugf("failed to delete veth %s in container: %v", containerLink.Attrs().Name, _err)
+				}
+			} else {
+				ec.Log.Debugf("Successfully deleted veth %s in container", containerLink.Attrs().Name)
+			}
+			return _err
 		})
```

Comment:
We don't need to delete the veth in the container, right? Returning `nil` will cause the container runtime to delete it. I see that we do delete it in the IPv4 version as well, but I think we actually don't need either. Might be better to not do it for v6 to confirm, then we can always refactor v4 if there's time later